packages/core/src/crypto/JwsService.ts

```diff
@@ -140,22 +149,58 @@ export class JwsService {
     }
 
     return {
+      ...options,
       alg: options.alg,
-      jwk: options.jwk,
+      jwk: options.jwk?.toJson(),
       kid: options.kid,
     }
   }
+
+  private async jwkFromProtectedHeader(agentContext: AgentContext, protectedHeader: JwsProtectedHeader): Promise<Jwk> {
+    if (protectedHeader.jwk && protectedHeader.kid) {
+      throw new AriesFrameworkError(
+        'Both JWK and kid are defined in the protected header. Only one of the two is allowed.'
+      )
+    }
+
+    // Jwk
+    if (protectedHeader.jwk) {
+      return getJwkFromJson(protectedHeader.jwk)
+    }
+
+    // Kid
+    if (protectedHeader.kid) {
+      if (!protectedHeader.kid.startsWith('did:')) {
+        throw new AriesFrameworkError(
+          `Only DIDs are supported as the 'kid' parameter for JWS. '${protectedHeader.kid}' is not a did.`
+        )
+      }
+
+      const didResolver = agentContext.dependencyManager.resolve(DidResolverService)
+      const didDocument = await didResolver.resolveDidDocument(agentContext, protectedHeader.kid)
+
+      // This is a special case for Aries RFC 0017 signed attachments. It allows a did:key without a keyId to be used kid
+      // https://github.com/hyperledger/aries-rfcs/blob/main/concepts/0017-attachments/README.md#signing-attachments
+      if (protectedHeader.kid.startsWith('did:key:') && !protectedHeader.kid.includes('#')) {
+        return getJwkFromKey(DidKey.fromDid(protectedHeader.kid).key)
+      }
+
+      // TODO: allowedPurposes
```

Comment:
What does this TODO mean? Would be quite difficult I think for someone else to pick this up.

Reply:
I'll extend it. It means that we should look at specific specific purposes from the did document (`authentication` vs `capabilityDelegation`, etc..).

But I think it's just authentication that we need here, so I'll pick that for now (better to be strict)